File: kbinding-annotations/src/main/java/com/benny/library/kbinding/internal/BindingInitializer.java
package com.benny.library.kbinding.internal;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by benny on 3/4/16.
 */

public class BindingInitializer {
    public static Map<Class<?>, BindingBuilder> initializers = new HashMap<>();

    @SuppressWarnings("unchecked")
    public static void init(Object obj) {
        getOrCreateBuilder(obj.getClass()).build(obj);
    }

    private static BindingBuilder getOrCreateBuilder(Class<?> clazzz) {
        BindingBuilder builder = initializers.get(clazzz);
        if(builder == null) {
            try {
                builder = (BindingBuilder) Class.forName(clazzz.getCanonicalName() + "$$KB").newInstance();
            }
            catch (Exception e) {
                builder = new MockBindingBuilder();
            }
            initializers.put(clazzz, builder);
        }
        return builder;
    }

    public static class MockBindingBuilder implements BindingBuilder<Object> {
        @Override
        public void build(Object target) {
        }
    }
}

File: kbinding/src/main/kotlin/com/benny/library/kbinding/bind/BindingInitializer.kt
package com.benny.library.kbinding.bind

/**
 * Created by benny on 3/4/16.
 */

object BindingInitializer {
    var builders: MutableMap<Class<*>, BindingBuilder<*>> = mutableMapOf()

    @SuppressWarnings("unchecked")
    fun init(obj: Any) {
        getOrCreateBuilder(obj.javaClass).build(obj)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> getOrCreateBuilder(clazz: Class<T>): BindingBuilder<T> {
        return builders.getOrPut(clazz, {
            try {
                return Class.forName(clazz.name + "\$\$KB").newInstance() as BindingBuilder<T>
            } catch (e: Exception) {
                return MockBindingBuilder()
            }
        }) as BindingBuilder<T>
    }

    class MockBindingBuilder<T> : BindingBuilder<T> {
        override fun build(target: T) {
        }
    }
}
